### PR TITLE
Feature/redis

### PR DESCRIPTION
**Pull request by jmassey-jacques**
I had trouble reaching 100% code coverage for lines, however, I noticed that the branch coverage is generally only around 50%, which my code is consistent with. Feedback on how I can test the missed lines would be appreciated.

### CODE REVIEW

**Comment by cinemast** [comment]
Thanks for raising this PR. Can you check if you can fix the build errors for Debian8 and Ubuntu1604?

File: src/examples/redisclient.cpp [hunk 1]
```diff
@@ -0,0 +1,36 @@
+/**
+ * @file simpleclient.cpp
+ * @date 03.01.2013
+ * @author Peter Spiess-Knafl <dev@spiessknafl.at>
```
Comment (cinemast) [inline src/examples/redisclient.cpp]: Please fix the header

File: src/jsonrpccpp/client/connectors/redisclient.cpp [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+ * @param len The length of the random string.
+ */
+void genRandom(char *s, int len) {
+    static const char alphanum[] =
+        "0123456789"
+        "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+        "abcdefghijklmnopqrstuvwxyz";
+
+    for (int i = 0; i < len; ++i) {
+        s[i] = alphanum[rand() % (sizeof(alphanum) - 1)];
```
Comment (cinemast) [inline src/jsonrpccpp/client/connectors/redisclient.cpp]: Doesn't rand need to be initialized somehow?

File: src/test/test_connector_redis.cpp [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Start Redis server in the background
+    //
+
+    /**
+     * This method will kill the redis server when we are done with it.
+     * @param pid The pid of the redis server to kill.
+     */
+    void stop_redis_server(int pid) {
```
Comment (cinemast) [inline src/test/test_connector_redis.cpp]: Could you move the initialisation of the redis binary to a separate class? e.g. `TestRedisServer`

File: src/test/test_connector_redis.cpp [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+{
+    RedisClient client2(TEST_HOST, TEST_PORT, "invalid_queue");
+    client2.SetTimeout(1);
+
+    string result;
+    CHECK_EXCEPTION_TYPE(client2.SendRPCMessage("examplerequest", result), JsonRpcException, check_exception1);
+}
+
+
+TEST_CASE_METHOD(F, "test_redis_server_longpost", TEST_MODULE)
```
Comment (cinemast) [inline src/test/test_connector_redis.cpp]: This is not a post request, please rename the test to `test_redis_server_longrequest`.

File: src/jsonrpccpp/client/connectors/redisclient.h [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+#include "../iclientconnector.h"
+#include <jsonrpccpp/common/exception.h>
+#include <hiredis/hiredis.h>
+
+namespace jsonrpc
+{
+    /**
+     * This class is the redis implementation of an AbstractClientConnector.
+     * It uses hiredis to connect to the redis server to send and receive with
+     * the RedisServer connector.
```
Comment (cinemast) [inline src/jsonrpccpp/client/connectors/redisclient.h]: Please also describe which redis commands are used and how the Queues are used. 
From what I saw the server always listens on a incoming queue and then responds to a separate response queue which is random for each client, right?

This should be mentioned in the comments.

File: src/jsonrpccpp/server/connectors/redisserver.h [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+#include <pthread.h>
+
+#include "../abstractserverconnector.h"
+
+namespace jsonrpc
+{
+  /**
+   * This class is the redis implementation of an AbstractServerConnector.
+   * It uses hiredis to connect to the redis server to performs its job.
+   * Each client request is handled in a new thread.
```
Comment (cinemast) [inline src/jsonrpccpp/server/connectors/redisserver.h]: Again, please describe the queue handling and used redis commands here.

File: cmake/libjsonrpccpp-common.pc.cmake [hunk 7]
```diff
@@ -1,5 +1,5 @@
 Name: libjsonrpccpp-common
 Description: Common libraries for libjson-rpc-cpp
 Version: ${MAJOR_VERSION}.${MINOR_VERSION}.${PATCH_VERSION}
-Libs: -L${FULL_PATH_LIBDIR} -ljsoncpp
+Libs: -L${FULL_PATH_LIBDIR} -ljsoncpp -lhiredis
```
Comment (cinemast) [inline cmake/libjsonrpccpp-common.pc.cmake]: But I guess the server and client .pc files need the -lhiredis flag.

**Comment by jmassey-jacques** [comment]
The issue appears to be a preexisting memory bug with filedescriptor connector. Investigating with valgrind at the moment.

File: cmake/libjsonrpccpp-common.pc.cmake (hunk 7 above)
Comment (jmassey-jacques) [inline cmake/libjsonrpccpp-common.pc.cmake]: Done.

File: src/examples/redisclient.cpp (hunk 1 above)
Comment (jmassey-jacques) [inline src/examples/redisclient.cpp]: Done.

File: src/jsonrpccpp/client/connectors/redisclient.cpp (hunk 2 above)
Comment (jmassey-jacques) [inline src/jsonrpccpp/client/connectors/redisclient.cpp]: Yes. With srand, which I use in the constructor for RedisClient.

File: src/test/test_connector_redis.cpp (hunk 4 above)
Comment (jmassey-jacques) [inline src/test/test_connector_redis.cpp]: Done.

File: src/jsonrpccpp/server/connectors/redisserver.h (hunk 6 above)
Comment (jmassey-jacques) [inline src/jsonrpccpp/server/connectors/redisserver.h]: Done.

File: src/jsonrpccpp/client/connectors/redisclient.h (hunk 5 above)
Comment (jmassey-jacques) [inline src/jsonrpccpp/client/connectors/redisclient.h]: Done.

File: src/test/test_connector_redis.cpp (hunk 3 above)
Comment (jmassey-jacques) [inline src/test/test_connector_redis.cpp]: Done.

**Comment by jmassey-jacques** [comment]
Pretty confused as to why debian8 is still failing. I just did a full travis build against the same commit on my repo and it all came up fine. Valgrind found quite a few issues so I'll go back to looking at that.

**Comment by cinemast** [comment]
Me too, apparently it is also failing now on develop. So I guess I'll have to take a look at it. Until this is fixed, i will hold this PR back.

**Comment by cinemast** [comment]
Also I could not find your changes regarding the review?

**Comment by jmassey-jacques** [comment]
Thanks for those fixes. The builds are passing and the only complaints in valgrind are for libjsoncpp.
I'm not sure what you mean about the changes regarding the review? They should all be there in commit 1612079.

**Comment by cinemast** [comment]
Ah sorry, that was my fault.

**Comment by cinemast** [comment]
About the valgrind errors in jsoncpp, I am already discussing this here: https://github.com/open-source-parsers/jsoncpp/issues/578

**Pull request merged**